### PR TITLE
Dynamic entry support

### PR DESCRIPTION
**Pull request by nkzawa**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**If relevant, link to documentation update:**

**Summary**
Enable to change the value of the `entry` option during watch mode.
Related: https://github.com/webpack/webpack/pull/3354, https://github.com/webpack/webpack/issues/2874

**Does this PR introduce a breaking change?**
No.

**Other information**

### CODE REVIEW

File: lib/MultiModule.js [hunk 1]
```diff
@@ -19,7 +20,9 @@ MultiModule.prototype = Object.create(Module.prototype);
 MultiModule.prototype.constructor = MultiModule;
 
 MultiModule.prototype.identifier = function() {
-	return "multi " + this.name;
+	return "multi " + this.dependencies.map(function(d) {
+		return path.resolve(this.context, d.request);
+	}.bind(this)).join("!");
```
Comment (sokra) [inline lib/MultiModule.js]: Why was this changed?

Comment (nkzawa) [inline lib/MultiModule.js]: This is because an entry with a certain name might has different dependencies when the entry function was called multiple times.

Comment (sokra) [inline lib/MultiModule.js]: Ok makes sense. Could you update readableIdentifier too and completely remove `this.name` from the MultiModule.

Comment (nkzawa) [inline lib/MultiModule.js]: How should I update `readableIdentifier` ? I thought it's for human then `this.name` makes sense ?

**Comment by nkzawa** [comment]
@sokra @TheLarkInn Let me know if there are anything else I should work on.

File: lib/MultiModule.js (hunk 1 above)
Comment (sokra) [inline lib/MultiModule.js]: `readableIdentifier` should reflect `identifier` but in a human readable way. You can use `requestShortener.shorten` to generate a readable identifier from a request.

---

Other issue: identifier should only use `d.request` and not `this.context`. And don't join them by `!`. This doesn't make sense here (`!` is for loaders). Maybe just join them by `|` or space.

**Comment by nkzawa** [comment]
@sokra Fixed `identifier` and `readableIdentifier`. Please check.

File: lib/MultiModule.js [hunk 2]
```diff
@@ -21,12 +20,14 @@ MultiModule.prototype.constructor = MultiModule;
 
 MultiModule.prototype.identifier = function() {
 	return "multi " + this.dependencies.map(function(d) {
-		return path.resolve(this.context, d.request);
-	}.bind(this)).join("!");
+		return d.request;
+	}.bind(this)).join(" ");
```
Comment (chicoxyzzy) [inline lib/MultiModule.js]: this should use second parameter of `map` or even better you can use an arrow function

File: lib/MultiModule.js [hunk 3]
```diff
@@ -21,12 +20,14 @@ MultiModule.prototype.constructor = MultiModule;
...
 };
 
-MultiModule.prototype.readableIdentifier = function() {
-	return "multi " + this.name;
+MultiModule.prototype.readableIdentifier = function(requestShortener) {
+	return "multi " + this.dependencies.map(function(d) {
+		return requestShortener.shorten(d.request);
+	}.bind(this)).join(" ");
```
Comment (chicoxyzzy) [inline lib/MultiModule.js]: same here

File: lib/MultiEntryPlugin.js [hunk 4]
```diff
@@ -12,6 +12,13 @@ function MultiEntryPlugin(context, entries, name) {
 	this.name = name;
 }
 module.exports = MultiEntryPlugin;
+MultiEntryPlugin.createDependency = function(entries, name) {
+	return new MultiEntryDependency(entries.map(function(e, idx) {
+		var dep = new SingleEntryDependency(e);
+		dep.loc = name + ":" + (100000 + idx);
+		return dep;
+	}, this), name);
```
Comment (chicoxyzzy) [inline lib/MultiEntryPlugin.js]: use arrow function here please

File: lib/MultiEntryPlugin.js [hunk 5]
```diff
@@ -22,10 +29,7 @@ MultiEntryPlugin.prototype.apply = function(compiler) {
...
 	});
 	compiler.plugin("make", function(compilation, callback) {
-		compilation.addEntry(this.context, new MultiEntryDependency(this.entries.map(function(e, idx) {
-			var dep = new SingleEntryDependency(e);
-			dep.loc = this.name + ":" + (100000 + idx);
-			return dep;
-		}, this), this.name), this.name, callback);
+		var dep = MultiEntryPlugin.createDependency(this.entries, this.name);
+		compilation.addEntry(this.context, dep, this.name, callback);
 	}.bind(this));
```
Comment (chicoxyzzy) [inline lib/MultiEntryPlugin.js]: use `compiler` instead of `this`, add arrow function and remove `bind`

File: lib/SingleEntryPlugin.js [hunk 6]
```diff
@@ -10,15 +10,19 @@ function SingleEntryPlugin(context, entry, name) {
...
 		var normalModuleFactory = params.normalModuleFactory;
 
 		compilation.dependencyFactories.set(SingleEntryDependency, normalModuleFactory);
 	});
 	compiler.plugin("make", function(compilation, callback) {
-		var dep = new SingleEntryDependency(this.entry);
-		dep.loc = this.name;
+		var dep = SingleEntryPlugin.createDependency(this.entry, this.name);
 		compilation.addEntry(this.context, dep, this.name, callback);
 	}.bind(this));
```
Comment (chicoxyzzy) [inline lib/SingleEntryPlugin.js]: same here

File: lib/MultiEntryPlugin.js (hunk 5 above)
Comment (nkzawa) [inline lib/MultiEntryPlugin.js]: I think fixing styles is not the scope of this PR. I will fix what I added.

Comment (chicoxyzzy) [inline lib/MultiEntryPlugin.js]: fair point

File: lib/DynamicEntryPlugin.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+			})
+		});
+	}
+}
+
+module.exports = DynamicEntryPlugin;
+
+DynamicEntryPlugin.createDependency = function(entry, name) {
+	if(Array.isArray(entry))
+		return new MultiEntryPlugin.createDependency(entry, name);
```
Comment (sokra) [inline lib/DynamicEntryPlugin.js]: The `new` is incorrect here.

Comment (sokra) [inline lib/DynamicEntryPlugin.js]: `new` is incorrect here.

**Comment by TheLarkInn** [comment]
Sync master and I think we are looking good.

**Comment by nkzawa** [comment]
@TheLarkInn fixed conflicts.

**Comment by TheLarkInn** [comment]
Sorry we've been merging a lot lately. Mind syncing again and I'll.merge right away.

**Comment by nkzawa** [comment]
@TheLarkInn done.

**Comment by TheLarkInn** [comment]
Thanks!!
